src/java.base/share/classes/jdk/internal/foreign/GlobalSession.java

```diff
@@ -67,4 +67,10 @@ void addInternal(ResourceList.ResourceCleanup resource) {
     public void justClose() {
         throw nonCloseable();
     }
+
+    @Override
+    @ForceInline
+    public void checkValidStateRaw() {
+        // do nothing, avoid liveness check
```

Comment:
This looks ok, although (see comments on MemorySessionImpl::checkValidStateRaw) sometimes having too many versions of this method can cause profile pollution issues (which is why we consolidated the impl at some point). I suggest trying other benchmarks, such as LoopOverPollutedXYZ

Reply:
Good point. Something I also realized over the weekend. I had thought bimorphic inlining would save us (and we overload other methods such as acquire/release already). But, we have 3 concrete implementation classes of MemorySessionImpl: confined, shared, and global. This blows up inlining it seems, causes a big regression when all three session kinds are used (see latest push I did):

```
Before:

Benchmark                                              Mode  Cnt  Score    Error  Units
LoopOverPollutedSegments.heap_segment_floats_VH        avgt   30  0.311 ±  0.001  ms/op
LoopOverPollutedSegments.heap_segment_floats_instance  avgt   30  0.311 ±  0.001  ms/op
LoopOverPollutedSegments.heap_segment_ints_VH          avgt   30  0.311 ±  0.001  ms/op
LoopOverPollutedSegments.heap_segment_ints_instance    avgt   30  0.311 ±  0.001  ms/op
LoopOverPollutedSegments.heap_unsafe                   avgt   30  0.217 ±  0.001  ms/op
LoopOverPollutedSegments.native_segment_VH             avgt   30  0.237 ±  0.005  ms/op
LoopOverPollutedSegments.native_segment_instance       avgt   30  0.245 ±  0.005  ms/op
LoopOverPollutedSegments.native_unsafe                 avgt   30  0.221 ±  0.007  ms/op

After:

Benchmark                                              Mode  Cnt  Score    Error  Units
LoopOverPollutedSegments.heap_segment_floats_VH        avgt   30  3.310 ±  0.005  ms/op
LoopOverPollutedSegments.heap_segment_floats_instance  avgt   30  3.333 ±  0.004  ms/op
LoopOverPollutedSegments.heap_segment_ints_VH          avgt   30  3.305 ±  0.007  ms/op
LoopOverPollutedSegments.heap_segment_ints_instance    avgt   30  3.439 ±  0.106  ms/op
LoopOverPollutedSegments.heap_unsafe                   avgt   30  0.217 ±  0.001  ms/op
LoopOverPollutedSegments.native_segment_VH             avgt   30  3.595 ±  0.043  ms/op
LoopOverPollutedSegments.native_segment_instance       avgt   30  3.696 ±  0.006  ms/op
LoopOverPollutedSegments.native_unsafe                 avgt   30  0.204 ±  0.001  ms/op
```

Reply:
I've also tried another solution which puts an `instanceof` check in the existing `checkValidStateRaw`. This does improve the case with the special `ALL` segment, but it also regresses the polluted case by about just as much it seems.

I think this is a dead end.

Reply:
IMHO, the only sane way to add support for this is to have a well-known constant - e.g. MemorySegment::allNative (which should be a restricted factory of course). Then special case this constant w.r.t. bound checks and temporal checks (since there's a well-known constant to compare against, I'm confident that pollution will not rear its ugly head). But that brings in a lot of API questions: for instance, if this segment is unbounded, what happens if you slice (of course one possible answer might be to throw). While we *might* add something like that in the future, I don't think we have enough evidence yet that something like this is really required.

Reply:
I suppose that is not too much different from the instanceof check I tried. The issue I see is that, while when the segment/scope are constant, everything folds away, everybody else now has to do an additional identity check on every access.

Reply:
> So what's the next step?

I believe the next best step is to try and determine exactly where the 2-3% performance regression comes from. We have many benchmarks that show no extra cost compared with unsafe if access occurs in a counted loop (because most we made sure that all checks can be hoisted outside the loop). But we need to be able to isolate the access pattern that is causing issue: does your code have a lot of "lone" reads/writes scattered through the code not surrounded by any loop? Or, maybe the loop is there, but for some reason the JIT is failing to hoist some of the checks? Or maybe a failure to vectorize the loop accordingly? These are very different problems - with the latter two probably more fixable (either with some changes in your code, on in our VM optimization logic) than the first (a **single** memory access using MemorySegment, ByteBuffer or any other API which enforces some extra checks is always going to be slower than same access using Unsafe). But loop-access is, typically, where the big wins are performance-wise, with potential for check hoisting, vectorization, and more.

Reply:
There's a ton of different access patterns, and most of them cannot be implemented as simple loops. This statement concerns me the most: "a single memory access using MemorySegment, ByteBuffer or any other API which enforces some extra checks is always going to be slower than same access using Unsafe"  If the MemorySegment is global and encompasses the entire address space, why should there be any extra checks at all compared to Unsafe?

Reply:
> There's a ton of different access patterns, and most of them cannot be implemented as simple loops. This statement concerns me the most: "a single memory access using MemorySegment, ByteBuffer or any other API which enforces some extra checks is always going to be slower than same access using Unsafe" If the MemorySegment is global and encompasses the entire address space, why should there be any extra checks at all compared to Unsafe?

The main point I'm trying to make here is that, so far at least, we're speculating as to what the cause of the regression might be. What we need to do is (a) be able to reproduce the issue you are experiencing (there is no evidence so far that the fix in this PR does anything to ameliorate your situation and I wouldn't be too surprised if it did not help at all) and (b) find some ways to mitigate it (where and if possible). For (a) we need help, as we are essentially discussing code without being able to see it, which is not optimal.

Reply:
Stuff like this can be very helpful, thanks!

Reply:
Panama vs. Unsafe: https://gist.github.com/broneill/3a39051635e3cb758d0cca5a963c685e

Reply:
Thanks for the benchmark. We've been able to reproduce your result. We are investigating further (that will take some time).

Reply:
Filed this:
https://bugs.openjdk.org/browse/JDK-8311932